Finn-Cullen-Onboarding-Challenge-Submission

FATCullen opened the pull request:
# Purpose
Verify solutions to onboarding challenge
# New Changes
Added thermal data reading and management
# Testing
Ran unit and integration tests locally
# Outstanding Changes

FATCullen committed:
Thermal data reading

FATCullen committed:
Implemented thermal data management

FATCullen committed:
Added extra error codes

Yarik-Popov commented on lm75bd/lm75bd.h:
```diff
@@ -5,7 +5,7 @@
 
 /* LM75BD I2C Device Address */
-#define LM75BD_OBC_I2C_ADDR /* Define the address here */
+#define LM75BD_OBC_I2C_ADDR 0b1001111
```
Use hex as thats what we use on the team

Yarik-Popov commented on services/thermal_mgr/thermal_mgr.c:
```diff
@@ -42,19 +43,40 @@ void initThermalSystemManager(lm75bd_config_t *config) {
-  /* Send an event to the thermal manager queue */
+  if(event == NULL) return ERR_CODE_INVALID_ARG;
+
+  if(xQueueSend(thermalMgrQueueHandle, event, 1) != pdPASS){
```
- [ ] check that the queue handle isn't null before use
- [ ] use block time of 0

Yarik-Popov commented on services/thermal_mgr/thermal_mgr.c:
```diff
@@ -42,19 +43,40 @@ void initThermalSystemManager(lm75bd_config_t *config) {
+    else if(temp <= LM75BD_DEFAULT_HYST_THRESH) safeOperatingConditions();
+    // Case neither
+    else LOG_ERROR_CODE(ERR_CODE_INVALID_INTERUPT);
+  }
```
Do not read in the isr, send an event using the function u wrote and have the thermalMgr task read and handle this logic

Yarik-Popov commented on lm75bd/lm75bd.c:
```diff
@@ -27,6 +27,29 @@ error_code_t lm75bdInit(lm75bd_config_t *config) {
+//  *temp = convertedTemp * 0.125;
+
+//  Conversion from twos complement, and conversion to Celsius
+  uint16_t convertedTemp = (readBuffer[0] << 3) | (readBuffer[1] >> 5);
+  if(readBuffer[0] & 0x80){
+      convertedTemp = (convertedTemp ^ 0x07FF) + 1;
+      *temp = convertedTemp * -0.125;
+  }
+  else{
+      *temp = convertedTemp * 0.125;
+  }
```
This code is overly complicated. You can avoid most of this code if you use a int16_t for the convertedTemp 
```suggestion 
int16_t convertedTemp=((readBuffer[0]<<8)|readBuffer[1])>>5;
*temp=convertedTemp*0.125;
```

Yarik-Popov commented on services/thermal_mgr/thermal_mgr.c:
```diff
@@ -42,19 +43,40 @@ void initThermalSystemManager(lm75bd_config_t *config) {
+    if(xQueueReceive(thermalMgrQueueHandle, &currentEvent, portMAX_DELAY) == pdPASS){
+      if(currentEvent.type == THERMAL_MGR_EVENT_MEASURE_TEMP_CMD){
+        float temp;
+        if(readTempLM75BD(LM75BD_OBC_I2C_ADDR, &temp) == ERR_CODE_SUCCESS) addTemperatureTelemetry(temp);
```
In general, always log the error code returned from non void functions. You should be logging the error code returned from the read temp

Yarik-Popov reviewed (changes requested):
Overall great start. Some changes are needed to improve the code quality

FATCullen committed:
Updated temp reading functions, simplified conversion to float

FATCullen committed:
Add files via upload

FATCullen committed:
Updated thermal data management

Now returns all non-success error codes, OS handler now sends event instead of handling interrupts itself

Yarik-Popov commented on services/thermal_mgr/thermal_mgr.c:
```diff
@@ -42,19 +43,52 @@ void initThermalSystemManager(lm75bd_config_t *config) {
+        //  Check that temp read is successful
+        if(errCode != ERR_CODE_SUCCESS) LOG_ERROR_CODE(errCode);
+        //  Check if over temp interrupt
+        else if(temp >= LM75BD_DEFAULT_OT_THRESH) overTemperatureDetected();
+        //  Check if hysteresis interrupt
+        else if(temp <= LM75BD_DEFAULT_HYST_THRESH) safeOperatingConditions();
+        //  Check if neither
+        else LOG_ERROR_CODE(ERR_CODE_INVALID_INTERUPT);
```
This should be under an else (or can put continue inside the if check) as you can't assume that temp will be any valid value.

Yarik-Popov replied:
Mb, you are right. Temp will be valid as the if case above handles that

Yarik-Popov reviewed (changes requested):
One last thing. Once this is handled, you are done the onboarding

Yarik-Popov commented:
Mb about the last comment. Lgtm, dm me your email and I'll add ou to the notion
